cmd/add.go

```diff
@@ -1,54 +1,155 @@
 // Copyright (c) 2023 Cisco Systems, Inc. and its affiliates
 // All rights reserved.
-
 package cmd
 
 import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
 	"github.com/cisco-open/grabit/internal"
+	"github.com/rs/zerolog/log"
 	"github.com/spf13/cobra"
 )
 
 func addAdd(cmd *cobra.Command) {
 	addCmd := &cobra.Command{
-		Use:   "add",
+		Use:   "add [url]",
 		Short: "Add new resource",
-		Args:  cobra.MinimumNArgs(1),
+		Args:  cobra.ExactArgs(1),
 		RunE:  runAdd,
 	}
+
 	addCmd.Flags().String("algo", internal.RecommendedAlgo, "Integrity algorithm")
 	addCmd.Flags().String("filename", "", "Target file name to use when downloading the resource")
 	addCmd.Flags().StringArray("tag", []string{}, "Resource tags")
+	addCmd.Flags().String("cache", "", "URL of Artifactory cache")
+
 	cmd.AddCommand(addCmd)
 }
 
 func runAdd(cmd *cobra.Command, args []string) error {
+	// Get flags
 	lockFile, err := cmd.Flags().GetString("lock-file")
 	if err != nil {
-		return err
-	}
-	lock, err := internal.NewLock(lockFile, true)
-	if err != nil {
-		return err
+		return fmt.Errorf("failed to get lock-file flag: %w", err)
 	}
+
 	algo, err := cmd.Flags().GetString("algo")
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get algo flag: %w", err)
 	}
+
 	tags, err := cmd.Flags().GetStringArray("tag")
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get tag flag: %w", err)
 	}
+
 	filename, err := cmd.Flags().GetString("filename")
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get filename flag: %w", err)
+	}
+
+	cache, err := cmd.Flags().GetString("cache")
+	if err != nil {
+		return fmt.Errorf("failed to get cache flag: %w", err)
+	}
+
+	// Validate cache configuration
+	if cache != "" {
+		// If cache URL is provided, authentication is required
+		token := os.Getenv("ARTIFACTORY_TOKEN")
+		if token == "" {
+			return fmt.Errorf("ARTIFACTORY_TOKEN must be set when using cache")
+		}
+		log.Debug().
+			Str("cache_url", cache).
+			Msg("Artifactory cache configured")
+	}
+
+	// Create or open lock file
+	lock, err := internal.NewLock(lockFile, true)
+	if err != nil {
+		return fmt.Errorf("failed to create/open lock file: %w", err)
+	}
+
+	// Check if resource is local file
+	isLocal := !strings.HasPrefix(args[0], "http://") && !strings.HasPrefix(args[0], "https://")
+	if isLocal && cache != "" {
+		log.Debug().
+			Str("file", args[0]).
+			Str("cache", cache).
+			Msg("Uploading local file to Artifactory")
+
+		if err := uploadToArtifactory(args[0], cache); err != nil {
```

Comment:
I don't think the file has actually been downloaded yet, has it?

Also, this is an interesting meta question -- I don't think we talked about this case last week: during `grabit add --cache ...`, should grabit upload the file to the cache?  Or does grabit upload the file to the cache the first time you run `grabit download`?